librz/include/rz_arch.h

```diff
@@ -59,6 +61,245 @@ RZ_API bool rz_arch_load_platform_sdb(RZ_NONNULL RzArchPlatformTarget *t, RZ_NON
+	RZ_ARCH_OPCODE_TYPE_ATOMIC /*                */ = (1ull << 44), ///< Opcode performs an atomic operation, to not be confused with memory fence/barrier
+	RZ_ARCH_OPCODE_TYPE_CONDITIONAL /*           */ = (1ull << 45), ///< Opcode performs an conditional operation
+	RZ_ARCH_OPCODE_TYPE_INDIRECT /*              */ = (1ull << 46), ///< Opcode performs an indirect jump/call/return operation
+	RZ_ARCH_OPCODE_TYPE_REGISTER /*              */ = (1ull << 46), ///< Opcode performs an register jump/call/return operation
+	RZ_ARCH_OPCODE_TYPE_MEMORY /*                */ = (1ull << 47), ///< Opcode performs an memory jump/call/return operation
+	RZ_ARCH_OPCODE_TYPE_IO /*                    */ = (1ull << 48), ///< Opcode performs an I/O operation
+	RZ_ARCH_OPCODE_TYPE_MMIO /*                  */ = (1ull << 49), ///< Opcode performs an memory I/O operation
+	RZ_ARCH_OPCODE_TYPE_REPEAT /*                */ = (1ull << 50), ///< Opcode performs a repeat N-times operation
+
+	/* Extended Jump Instructions */
+	RZ_ARCH_OPCODE_TYPE_REGISTER_JUMP = (RZ_ARCH_OPCODE_TYPE_JUMP_UNKNOWN | RZ_ARCH_OPCODE_TYPE_REGISTER),
+	RZ_ARCH_OPCODE_TYPE_INDIRECT_JUMP = (RZ_ARCH_OPCODE_TYPE_JUMP_UNKNOWN | RZ_ARCH_OPCODE_TYPE_INDIRECT),
+	RZ_ARCH_OPCODE_TYPE_INDIRECT_REGISTER_JUMP = (RZ_ARCH_OPCODE_TYPE_INDIRECT_JUMP | RZ_ARCH_OPCODE_TYPE_REGISTER),
+
+	/* Extended Conditional Jump Instructions */
+	RZ_ARCH_OPCODE_TYPE_CONDITIONAL_JUMP = (RZ_ARCH_OPCODE_TYPE_JUMP | RZ_ARCH_OPCODE_TYPE_CONDITIONAL),
+	RZ_ARCH_OPCODE_TYPE_CONDITIONAL_JUMP_UNKNOWN = (RZ_ARCH_OPCODE_TYPE_JUMP_UNKNOWN | RZ_ARCH_OPCODE_TYPE_CONDITIONAL),
+	RZ_ARCH_OPCODE_TYPE_CONDITIONAL_REGISTER_JUMP = (RZ_ARCH_OPCODE_TYPE_CONDITIONAL_JUMP_UNKNOWN | RZ_ARCH_OPCODE_TYPE_REGISTER),
+
+	/* Extended Memory Jump Instructions */
+	RZ_ARCH_OPCODE_TYPE_MEMORY_JUMP = (RZ_ARCH_OPCODE_TYPE_JUMP | RZ_ARCH_OPCODE_TYPE_MEMORY),
+	RZ_ARCH_OPCODE_TYPE_MEMORY_CONDITIONAL_JUMP = (RZ_ARCH_OPCODE_TYPE_CONDITIONAL_JUMP | RZ_ARCH_OPCODE_TYPE_MEMORY),
+
+	/* Extended Call/Return Instructions */
+	RZ_ARCH_OPCODE_TYPE_REGISTER_CALL = (RZ_ARCH_OPCODE_TYPE_CALL_UNKNOWN | RZ_ARCH_OPCODE_TYPE_REGISTER),
+	RZ_ARCH_OPCODE_TYPE_INDIRECT_CALL = (RZ_ARCH_OPCODE_TYPE_CALL_UNKNOWN | RZ_ARCH_OPCODE_TYPE_INDIRECT),
+	RZ_ARCH_OPCODE_TYPE_INDIRECT_REGISTER_CALL = (RZ_ARCH_OPCODE_TYPE_INDIRECT_CALL | RZ_ARCH_OPCODE_TYPE_REGISTER),
+	RZ_ARCH_OPCODE_TYPE_CONDITIONAL_CALL = (RZ_ARCH_OPCODE_TYPE_CALL | RZ_ARCH_OPCODE_TYPE_CONDITIONAL),
+	RZ_ARCH_OPCODE_TYPE_CONDITIONAL_CALL_UNKNOWN = (RZ_ARCH_OPCODE_TYPE_CALL_UNKNOWN | RZ_ARCH_OPCODE_TYPE_CONDITIONAL),
+	RZ_ARCH_OPCODE_TYPE_CONDITIONAL_RETURN = (RZ_ARCH_OPCODE_TYPE_RETURN | RZ_ARCH_OPCODE_TYPE_CONDITIONAL),
+} RzArchOpcodeType;
+
+typedef struct rz_arch_opcode_asm_t {
+	ut32 size; ///< opcode size in bytes
+	RzStrBuf buffer; ///< inline assembly string
+} RzArchOpcodeAsm;
```

Comment:
In Hexagon one instruction word (instruction packet) can encode multiple instructions. Although `instruction word != opcode` they are treated similar in practice (single address for instruction word, executed atomically, execution commits or fails, never partially commits).
It would be nice to account for that here. Maybe a flag (`multiple_opcodes = <bool>`) and `buffer` can contain new lines (each line an opcode assembly string)

Reply:
Just recognized that this will interfere with the opcode type flags. Maybe a `instruction` struct could be added which holds x opcodes?

Reply:
There is an additional problem of missing context.
Some instruction words only disassemble correctly if the previous instruction is valid.
Or when they are at a symbol offset.

Currently bytes at an address get decoded multiple times before anything is done (by asm, analysis, for emulation stepping etc.).
Each time it is possible that, with the exception of the instruction bytes at the address, everything else changes (e.g. buffer size, content of the buffer after the instruction bytes and their content (could be 0x00..., 0xff... or the next instruction bytes, no one knows!1!!) etc.).

Because the context changes, not even buffering helps reliably.

I think it is important that the passed data to the plugin is at a minimum nicely defined AND stays the same over multiple decoding requests for a single address.
Simply so the plugin doesn't get a random set of stuff to work with, whenever it is called at the same address.

I struggle so enormously often with this currently, that it would be a huge relive of pain.